### PR TITLE
Option to hide autogenerated protobuf fields

### PR DESCRIPTION
**Pull request by sirkon**
Hi!

We are using gRPC (with protobuf) and the latest version of `protoc-gen-go` generates 3 service fields named

```
XXX_NoUnkeyedLiteral
XXX_unrecognized
XXX_sizecache
```

and it is getting annoying to see them in the output. So, I added special option to hide these fields which is set to true by default.

I decided to filter out exactly these three fields instead of filtering out everything started from `XXX_` as explicit is better than implicit.

### CODE REVIEW

**Comment by atombender** [comment]
Hi, thanks for the pull request! This is rather specific to Protobuf. Similar filtering might be desirable in the future for other kinds of libraries. I propose that rather than making it specific to Protobuf, we add a simple filter function to the config, e.g.:

```go
type Config struct {
  // ...
  Filter func(name string) bool
}
```

Then you can do:

```go
var cfg litter.Config
cfg.Filter = func(name string) bool {
  return !strings.HasPrefix(name, "XXX_")
}
```

**Comment by sirkon** [comment]
Won't it be too hard in general case? Option approach let to control field filtering easily – to say, we can filter out Protobuf fields and show fields after another code generator. With `Filter` we need to write function for each specific case.

**Comment by atombender** [comment]
Well, how many such filters would you need? It seems like overkill. This is the first instance where filtering has come up at all, so it doesn't sound like a very common case.

If the goal is just to filter by name, you could just have:

```go
type Config struct {
  // ...
  FieldExclusions *regexp.Regexp
}
```

Then you can do:

```
var cfg litter.Config
cfg.FieldExclusions = regexp.MustCompile("^XXX_")
```

Simple is better.

**Comment by sirkon** [comment]
Regexp based filterer seems like the best idea, done. Check it please.

File: dump.go [hunk 1]
```diff
@@ -34,6 +35,7 @@ type Options struct {
 var Config = Options{
 	StripPackageNames: false,
 	HidePrivateFields: true,
+	FieldExclusions:   regexp.MustCompile(`^(XXX_.*)$`),
```
Comment (atombender) [inline dump.go]: Doing this by default is probably uncontroversial, but it should have a comment describing what `XXX_` is.

File: dump_test.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ import (
 	"os/exec"
 	"testing"
 
-	"github.com/sanity-io/litter"
+	"github.com/sirkon/litter"
```
Comment (atombender) [inline dump_test.go]: This is your fork.

File: README.md [hunk 3]
```diff
@@ -131,6 +131,9 @@ litter.Config.StripPackageNames = true
 // Hide private struct fields from dumped structs
 litter.Config.HidePrivateFields = true
 
+// Hide fields generated by `protoc-gen-go`
+litter.Config.HideProtobufFields = true
```
Comment (atombender) [inline README.md]: Should no longer be part of readme.

File: dump_test.go (hunk 2 above)
Comment (sirkon) [reply dump_test.go]: fixed

File: README.md (hunk 3 above)
Comment (sirkon) [reply README.md]: fixed too

File: dump.go (hunk 1 above)
Comment (sirkon) [reply dump.go]: Indeed, comment added

**Pull request merged**